### ISSUE TITLE
muxer and aac data problem

### ISSUE THREAD

**Issue by FunnyDevs** [pedroSG94/RootEncoder#1063]
Hello pedro,
i'm using your library and i would like to change the default muxer (editing code in RecordController, in `private void write(int track, ByteBuffer byteBuffer, MediaCodec.BufferInfo info`) .  
I have tried with jcodec and with ffmpeg, but i have  always problem with aac data: the mp4 output file has audio plying too fast than video and i don't understand why. Should it do some kind of editing on aac bytebuffer (except adding adts header that i do)?
Thank you.

**Comment by pedroSG94** [comment]
Probably your problem is related with timestamp. Make sure that your timestamp is correct.
You can get it from info.presentationTimeUs in micro seconds.
While you are recording, video and audio should have a similar timestamp to sync it or you will have that result.

**Comment by FunnyDevs** [comment]
How could i create a timestamp? Only with `System.currentTimeInMillis()`?
Because i have problems using the presentationTimeUs from bufferInfo in RecordController.java.

**Comment by pedroSG94** [comment]
First of all this code:
```java
audioPts = 1024 * audioframeNo;
```
Why this timestamp? 
Anyway, you can re use my timestamp like this:
```java
audioPts = bufferInfo.presentationTimeUs;
```
This use microseconds so maybe you need change it to milliseconds or other thing depend of your library used.

If you still want to create your own timestamp i'm doing it like this:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/video/VideoEncoder.java#L454
Where presentTimeUs is System.nanoTime() / 1000 when you start encoders.

**Comment by FunnyDevs** [comment]
using (only for audio)

`bufferInfo.presentationTimeUs`

improves the situation, but it seems (editing with vlc audio sync) the audio is behind 300 milliseconds. 
Debugging with android studio, the audio data arrives before the video data, so i had an insight: 
adding this control in recordAudio

```
public void recordAudio(ByteBuffer audioBuffer, MediaCodec.BufferInfo audioInfo) {
    if (status == Status.RECORDING) {
      if (videoInfo.flags == MediaCodec.BUFFER_FLAG_KEY_FRAME)
        return;

      updateFormat(this.audioInfo, audioInfo);
      write(1, audioBuffer, this.audioInfo);
    }
  }
```

 and now the sync if perfect!

**Comment by pedroSG94** [comment]
This could be interesting if I can handle to do it without add it directly to the main project.
The reason is that I don't want add a 3rd party library that could be abandoned in anytime and I can't solve issues if I have any problem.
If you can create a PR or share me code used I can try create a new project that could be used like a extra gradle that allow you use it similar to this:
https://github.com/pedroSG94/RTSP-Server
Basically a project that extend from this one and add other feature.

**Comment by FunnyDevs** [comment]
ok, but i think it should be a mix of what you said:
1)  PR to edit the RecordController class, adding two classes (interface Muxer.java and AndroidMediaMuxer with the default MediaMuxer ) and editing the methods of "startRecord" in  CameraBase classes
2) create a new project  with different implementation for Muxer.java (as Jcodec and Xuggler i have created)

**Comment by pedroSG94** [comment]
Closing as inactive.
